File: rosetta-test-py/url-parsing.py
import urllib.parse
import rosetta
import urllib 

suite = rosetta.suite("stdlib urlparsing", "rosetta-test-suites/url-parsing-rfc.ros")

@suite.placeholder("url-parse")
def url_parse(env, url_string):
    try:
        return urllib.parse.urlparse(url_string)
    except ValueError as e:
        return e

@suite.placeholder("parse-error?")
def parse_error(env, parse_result):
    return isinstance(parse_result, Exception)

@suite.placeholder("url-scheme")
def url_scheme(env, parse_result: urllib.parse.ParseResult):
    return parse_result.scheme or ""

@suite.placeholder("url-authority")
def url_authority(env, parse_result: urllib.parse.ParseResult):
    return parse_result.netloc or ""

@suite.placeholder("url-path")
def url_path(env, parse_result: urllib.parse.ParseResult):
    return parse_result.path or ""

#
# Running
#

suite.run()

File: rosetta-test-rb/url_parsing.rb
require "uri"
require_relative "rosetta"

suite "rosetta-test-suites/url-parsing-rfc.ros" do
  expected_failures :test_scheme_with_invalid_characters,
    :"test_non-terminated_scheme",
    :test_invalid_ipv4_host

  placeholder "url-parse" do |_env, url_string|
    URI.parse(url_string)
  rescue URI::InvalidURIError => e
    e
  end

  placeholder "parse-error?" do |_env, parse_result|
    parse_result.is_a?(URI::InvalidURIError)
  end

  placeholder "url-scheme" do |_env, uri|
    uri.scheme
  end

  placeholder "url-authority" do |_env, uri|
    authority = uri.host
    authority = "#{uri.userinfo}@#{authority}" if uri.userinfo
    authority = "#{authority}:#{uri.port}" if uri.port
    authority
  end
end
